File: cpp/questions/isValidBST.cpp
#include <bits/stdc++.h>
#include <functional>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    bool isValidBST(TreeNode* root) {
        long long pre = LONG_LONG_MIN;
        std::function<bool(TreeNode*)> io = [&](TreeNode* n) -> bool {                
            if (n == nullptr) {
                return true;
            }
            if (!io(n->left)) {
                return false;
            }
            if (n->val <= pre) {
                return false;
            }
            pre = n->val;
            return io(n->right);
        };
        return io(root);
    }
};

File: python/questions/isValidBST.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        pre = -float('inf')
        def io(n: Optional[TreeNode]) -> bool:
            nonlocal pre
            if n is None:
                return True
            if not io(n.left):
                return False
            if n.val <= pre:
                return False
            pre = n.val
            return io(n.right)
        return io(root)
